apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/TransactionManager.java

```diff
@@ -0,0 +1,193 @@
+package gov.cms.bfd.pipeline.sharedutils;
+
+import com.google.common.annotations.VisibleForTesting;
+import gov.cms.bfd.sharedutils.interfaces.ThrowingConsumer;
+import gov.cms.bfd.sharedutils.interfaces.ThrowingFunction;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+
+/**
+ * Manages life cycle of an {@link EntityManager} to execute one or more transactions. Executes as
+ * many transactions as possible with a single {@link EntityManager} but obtains a new one if any
+ * transaction throws an exception.
+ *
+ * <p>Takes care of the tedious process of ensuring that any exception thrown by the transaction
+ * itself does not get lost if the subsequent rollback or close operations also throw.
+ *
+ * <p>Instances are intended to be dedicated to a particular thread such as a worker thread from a
+ * thread pool rather than shared by many instances (although that is possible). Any given instance
+ * will only allow one thread at a time to execute a transaction.
+ */
+public class TransactionManager implements AutoCloseable {
+  /** Used to create new {@link EntityManager}s as needed. */
+  private final EntityManagerFactory entityManagerFactory;
+  /**
+   * The current {@link EntityManager}. May be null if we have not yet created one or if it was
+   * closed after a failed transaction.
+   */
+  @Nullable private EntityManager entityManager;
+
+  /**
+   * Creates a new instance using the provided {@link EntityManagerFactory}.
+   *
+   * @param entityManagerFactory the {@link EntityManagerFactory}
+   */
+  public TransactionManager(EntityManagerFactory entityManagerFactory) {
+    this.entityManagerFactory = entityManagerFactory;
+  }
+
+  /**
+   * Runs a transaction that produces a result (usually a query). Ensures that the transaction is
+   * properly committed if the function completes normally or rolled back if the function throws an
+   * exception.
+   *
+   * @param functionLogic logic to invoke with the {@link EntityManager}
+   * @param <T> return type of the function
+   * @param <E> exception type thrown by the function
+   * @throws E pass through exception from calling the function
+   * @return return value from the function
+   */
+  public synchronized <T, E extends Exception> T executeFunction(
+      ThrowingFunction<T, EntityManager, E> functionLogic) throws E {
+    Exception transactionException = null;
+    final var entityManager = getOrCreateEntityManager();
```

Comment:
Curious; can't createEntityManager trigger an exception as well.

Reply:
Good point.  It shouldn't cause any problems if it did though since in that case the exception would pass through to the caller and the `entityManager` field would remain null.